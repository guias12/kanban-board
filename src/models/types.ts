export interface ICard {
  id: number;
  text: string;
}

export interface IList {
  title: string;
  id: number;
  cards: ICard[];
}
