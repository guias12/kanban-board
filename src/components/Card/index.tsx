import React, { useState, KeyboardEvent } from 'react';
import { changeCardText } from '../../reducers/listSlice';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import { faSave, faTimes } from '@fortawesome/free-solid-svg-icons';
import { ICard } from '../../models/types';
import './styles.scoped.scss';
import { Draggable } from 'react-beautiful-dnd';
import { useDispatch } from 'react-redux';

interface ICardProps {
  card: ICard;
  id: number;
  index: number;
  listId: number;
}
const Card = ({ card, id, index, listId }: ICardProps) => {
  const dispatch = useDispatch();
  const [cardText, setCardText] = useState<string>(card.text);
  const [isEdittingCard, setIsEdditingCard] = useState<boolean>(false);

  const toggleIsEdditingCard = (): void => {
    setIsEdditingCard(true);
  };

  const handleInputKeydow = (event: KeyboardEvent<HTMLInputElement>) => {
    if (event.key === 'Enter') {
      saveNewCardText();
    }
  };

  const saveNewCardText = (): void => {
    setIsEdditingCard(false);
    if (cardText) {
      dispatch(
        changeCardText({
          listId: listId,
          text: cardText,
          cardId: id,
        })
      );
    }
  };

  const cancelEdittingMode = (): void => {
    setIsEdditingCard(false);
    setCardText(card.text);
  };
  return (
    <Draggable draggableId={`${id}`} index={index}>
      {(provided) => (
        <div
          className="card-container"
          ref={provided.innerRef}
          {...provided.draggableProps}
          {...provided.dragHandleProps}
        >
          {isEdittingCard ? (
            <div className="card-input-container">
              <input
                type="text"
                value={cardText}
                onChange={(e) => setCardText(e.target.value)}
                onKeyDown={handleInputKeydow}
                autoFocus
                className="card-text-input"
                data-testid="card-input"
              />
              <FontAwesomeIcon
                onClick={saveNewCardText}
                className="save-icon"
                data-testid="card-input-save"
                icon={faSave}
              />
              <FontAwesomeIcon
                onClick={cancelEdittingMode}
                className="cancel-icon"
                data-testid="card-input-cancel"
                icon={faTimes}
              />
            </div>
          ) : (
            <span
              className="card-text"
              onClick={toggleIsEdditingCard}
              data-testid="card-text"
            >
              {card.text}
            </span>
          )}
        </div>
      )}
    </Draggable>
  );
};

export default Card;
