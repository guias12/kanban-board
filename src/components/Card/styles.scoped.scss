.card-container {
  margin: 5px 0;
  background: #ffffff;
  box-shadow: rgba(149, 157, 165, 0.2) 0px 8px 24px;
  min-height: 25px;
  width: 90%;
  border-radius: 4px;
  text-align: center;
  cursor: grab;
  padding: 10px 0;
}

.card-text {
  cursor: text;
}
