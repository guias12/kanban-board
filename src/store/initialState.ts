import { IInitialState } from '../models/types';
const getInitialState = (): IInitialState => ({
  searchTerm: '',
  list: [
    {
      title: 'To Do',
      id: 0,
      cards: [
        {
          id: 0,
          text: 'Test kanban board',
        },
      ],
    },
    {
      title: 'Doing',
      id: 1,
      cards: [],
    },
    {
      title: 'Done',
      id: 2,
      cards: [],
    },
  ],
});

const getEhancedTestInitialState = (): IInitialState => ({
  searchTerm: '',
  list: [
    {
      title: 'To Do',
      id: 0,
      cards: [
        {
          id: 0,
          text: 'First ToDo card',
        },
        {
          id: 1,
          text: 'Second ToDo card',
        },
      ],
    },
    {
      title: 'Doing',
      id: 1,
      cards: [
        {
          id: 2,
          text: 'first Doing card',
        },
        {
          id: 3,
          text: 'second Doing card',
        },
        {
          id: 4,
          text: 'third Doing card',
        },
      ],
    },
    {
      title: 'Done',
      id: 2,
      cards: [
        {
          id: 5,
          text: 'first Done card',
        },
        {
          id: 6,
          text: 'second Done card',
        },
      ],
    },
  ],
});

export { getInitialState, getEhancedTestInitialState };
