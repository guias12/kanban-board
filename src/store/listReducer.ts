import { IInitialState } from '../models/types';
import { AnyAction } from 'redux';
import { CONSTANTS } from '../actions';

const initialState: IInitialState = {
  searchTerm: '',
  list: [
    {
      title: 'To Do',
      id: 0,
      cards: [
        {
          id: 0,
          text: 'Test kanban board',
        },
      ],
    },
    {
      title: 'Doing',
      id: 1,
      cards: [],
    },
    {
      title: 'Done',
      id: 2,
      cards: [],
    },
  ],
};
let nextAvailableCardId = 1;

const listReducer = (state = initialState, action: AnyAction) => {
  const stateClone: IInitialState = { ...state };

  switch (action.type) {
    case CONSTANTS.ADD_CARD:
      const newStateList = stateClone.list.map((list) => {
        if (list.id === action.payload.listId) {
          const newCardId = nextAvailableCardId;
          const newCard = {
            text: action.payload.text,
            id: newCardId,
          };
          nextAvailableCardId += 1;

          return {
            ...list,
            cards: [...list.cards, newCard],
          };
        } else {
          return list;
        }
      });

      stateClone.list = newStateList;
      state.list = newStateList;
      return stateClone;
    case CONSTANTS.CHANGE_CARD_TEXT:
      const { listId, cardId, text } = action.payload;
      let currentList = stateClone.list.find((list) => list.id === listId);
      const currentListIndex = stateClone.list.findIndex(
        (list) => list.id === listId
      );
      const edditingCardIndex = currentList?.cards.findIndex(
        (card) => card.id === cardId
      );

      if (
        (edditingCardIndex || edditingCardIndex === 0) &&
        (currentList || currentList === 0)
      ) {
        let newCards = [...currentList.cards];
        newCards[edditingCardIndex].text = text;
        currentList = {
          ...currentList,
          cards: [...newCards],
        };

        stateClone.list[currentListIndex] = currentList;
        return stateClone;
      }

      return state;
    case CONSTANTS.DRAG_CARD:
      const {
        sourceListId,
        destinationListId,
        sourceCardIndex,
        destinationCardIndex,
      } = action.payload;

      if (sourceListId === destinationListId) {
        const list = stateClone.list.find(
          (list) => sourceListId === String(list.id)
        );
        const card = list?.cards.splice(sourceCardIndex, 1);

        if (card) {
          list?.cards.splice(destinationCardIndex, 0, ...card);
        }
      } else {
        const sourceList = stateClone.list.find(
          (list) => sourceListId === String(list.id)
        );
        const card = sourceList?.cards.splice(sourceCardIndex, 1);
        const destinationList = stateClone.list.find(
          (list) => destinationListId === String(list.id)
        );
        if (card) {
          destinationList?.cards.splice(destinationCardIndex, 0, ...card);
        }
      }

      return stateClone;
    case CONSTANTS.FILTER_CARD:
      const { searchTerm } = action.payload;
      stateClone.searchTerm = searchTerm;

      return stateClone;
    default:
      return state;
  }
};

export default listReducer;
