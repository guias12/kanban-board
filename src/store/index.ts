import { configureStore } from '@reduxjs/toolkit';
import { TypedUseSelectorHook, useSelector } from 'react-redux';
import listReducer from './listSlice';

const store = configureStore({
  reducer: {
    list: listReducer,
  },
});

export type RootState = ReturnType<typeof store.getState>;

export type AppDispatch = typeof store.dispatch;

export const useAppSelector: TypedUseSelectorHook<RootState> = useSelector;

export default store;
