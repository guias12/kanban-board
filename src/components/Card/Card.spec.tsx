import React from 'react';
import Card from './index';
import { Droppable } from 'react-beautiful-dnd';
import { DragDropContext } from 'react-beautiful-dnd';
import { ICard } from '../../models/types';
import { render, screen } from '@testing-library/react';
import configureMockStore, { MockStoreEnhanced } from 'redux-mock-store';
import { getInitialState } from '../../store/initialState';
import { create } from 'react-test-renderer';
import { Provider } from 'react-redux';
import fireEvent from '@testing-library/user-event';

describe('Card component', () => {
  let store: MockStoreEnhanced<unknown, {}>;
  const setup = (): JSX.Element => {
    const initialState = getInitialState();
    store = configureMockStore()({ list: initialState });

    const card: ICard = {
      id: 0,
      text: 'test card',
    };

    return (
      <Provider store={store}>
        <DragDropContext onDragEnd={() => {}}>
          <Droppable droppableId={`0}`}>
            {(provided) => (
              <div {...provided.droppableProps} ref={provided.innerRef}>
                <Card card={card} id={card.id} index={0} listId={0} />
              </div>
            )}
          </Droppable>
        </DragDropContext>
      </Provider>
    );
  };

  it('Should render Card component correctly', () => {
    render(setup());
    const cardText = screen.getByTestId('card-text');

    expect(cardText).toBeInTheDocument();
  });

  it('Should allow edit Card text inline', () => {
    render(setup());
    const cardText = screen.getByTestId('card-text');
    fireEvent.click(cardText);

    const cardInput = screen.getByTestId('card-input');
    expect(cardInput).toBeInTheDocument();
  });

  it('Should cancel inline edition of Card text on button click', () => {
    render(setup());
    const cardText = screen.getByTestId('card-text');
    fireEvent.click(cardText);
    const cardInput = screen.getByTestId('card-input');

    const cancelEditionButton = screen.getByTestId('card-input-cancel');
    fireEvent.click(cancelEditionButton);

    expect(cardInput).not.toBeInTheDocument();
  });

  it('Should save inline edition of Card text on button click', () => {
    render(setup());
    const cardText = screen.getByTestId('card-text');
    fireEvent.click(cardText);
    const cardInput = screen.getByTestId('card-input');

    const cancelEditionButton = screen.getByTestId('card-input-save');
    fireEvent.click(cancelEditionButton);

    expect(cardInput).not.toBeInTheDocument();
  });

  it('Should match Card snapshot', () => {
    const component = create(setup());
    const componentJson = component.toJSON();

    expect(componentJson).toMatchSnapshot();
  });
});
