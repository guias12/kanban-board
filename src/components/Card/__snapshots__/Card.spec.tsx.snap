// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Card component Should match Card snapshot 1`] = `
<div
  data-rbd-droppable-context-id="4"
  data-rbd-droppable-id="0}"
>
  <div
    aria-describedby="rbd-hidden-text-4-hidden-text-12"
    className="card-container"
    data-rbd-drag-handle-context-id="4"
    data-rbd-drag-handle-draggable-id="0"
    data-rbd-draggable-context-id="4"
    data-rbd-draggable-id="0"
    draggable={false}
    onDragStart={[Function]}
    onTransitionEnd={null}
    role="button"
    style={
      Object {
        "transform": null,
        "transition": null,
      }
    }
    tabIndex={0}
  >
    <span
      className="card-text"
      data-testid="card-text"
      onClick={[Function]}
    >
      test card
    </span>
  </div>
</div>
`;
